Use a flipped coordinate system in framebuffers to avoid manual flipping of texture coordinates

hrydgard opened the pull request:
Simplifies a lot of things and gets rid of a few shaders, although there are a few new differences between buffered and non-buffered, as we are effectively flipped again when drawing to the backbuffer directly, which we do in non-buffered.

This will help #8016 .

hrydgard committed:
Remove a whole lot of flipping special cases

hrydgard committed:
Fix a few more flips

hrydgard committed:
Fix non-buffered rendering

hrydgard committed:
Move the viewport to the right place.

hrydgard committed:
Fix non-through rectangles

unknownbrackets commented on GPU/Directx9/TextureCacheDX9.cpp:
```diff
@@ -1012,7 +1011,7 @@ void TextureCacheDX9::ApplyTextureFramebuffer(TexCacheEntry *entry, VirtualFrame
 			const float uvleft = u1 * invWidth;
 			const float uvright = u2 * invWidth;
-			const float uvtop = 1.0f - v1 * invHeight;
+			const float uvtop = 1.0f - v1 * invHeight;  // TODO: Seems we should ditch the "1.0f - "
```
Yeah, definitely, the positions need to agree too.  Otherwise it won't use the correct subset.

-[Unknown]

unknownbrackets replied:
Oh, this is DX9... yeah maybe this is a typo, oops...

-[Unknown]

unknownbrackets commented on GPU/GLES/ShaderManager.cpp:
```diff
@@ -417,7 +419,7 @@ void LinkedShader::UpdateUniforms(u32 vertType) {
 			flippedMatrix[0] = -flippedMatrix[0];
 			flippedMatrix[12] = -flippedMatrix[12];
-		}
+		}\
```
Oops, heh.

-[Unknown]

hrydgard committed:
Shifted projection matrices need a few more parts flipped. This could fix bugs in games, too.

Fixes the strange camera issues in FF:CC.

hrydgard committed:
Remove more obsolete v-flip support

hrydgard committed:
Two wrongs can appently make a right. Remove an inadvertent flip, add flips where we actually need them.

unknownbrackets commented on GPU/GLES/Framebuffer.cpp:
```diff
@@ -659,7 +665,7 @@ void FramebufferManager::ResizeFramebufFBO(VirtualFramebuffer *vfb, u16 w, u16 h
 	vfb->fbo = fbo_create(vfb->renderWidth, vfb->renderHeight, 1, true, (FBOColorDepth)vfb->colorDepth);
 	if (old.fbo) {
-		INFO_LOG(SCEGE, "Resizing FBO for %08x : %i x %i x %i", vfb->fb_address, w, h, vfb->format);
+		WARN_LOG(SCEGE, "Resizing FBO for %08x : %i x %i x %i", vfb->fb_address, w, h, vfb->format);
```
Hmm, I don't think this is really a "warning" per se.

-[Unknown]

hrydgard replied:
Ah, that was supposed to be temporary.

LunaMoo commented:
Cool this fixed #4665

hrydgard commented:
Neat. Not quite sure why though :)

There are still a couple of bugs lurking so I can't merge yet.

hrydgard committed:
Minor cleanups

LunaMoo commented:
Well there are some problems with effects, where select cursor or maybe just all transparent effects, shows inverted graphics inside of them, but other than that looks nice:).
Possibly it also affected #6736 but I don't have this one.

unknownbrackets commented:
Okay, so just to take inventory:

What this does is draw everything "right side up" until drawing to the actual backbuffer.  Previously we used the OpenGL "upside down" coordinate system throughout all drawing.  Example:

```
4 XXXXXXXXXX      AAAAAAAAAA 0
3 XXXXXXXXXX      BBBBBBBBBB 1
2 CCCCCCCCCC  ->  CCCCCCCCCC 2
1 BBBBBBBBBB      XXXXXXXXXX 3
0 AAAAAAAAAA      XXXXXXXXXX 4
```

Then when drawing to the final backbuffer, the image is flipped appropriately to match the OpenGL system.  In this way, textures (which use the "right side up" coordinate system) match framebuffers.

Problems and what I assume the solution is (still need to audit, though I've reviewed the code):

<dl>
   <dt>Rendering coordinates</dt><dd>Flipped using the projection matrix</dd>
   <dt>Viewport params</dt><dd>No longer need to be flipped</dd>
   <dt>Texturing from framebuffers</dt><dd>No longer need to be flipped</dd>
   <dt>Memory -> framebuffer upload<dt><dd>Textured into "right side up" coordinate system</dd>
   <dt>Framebuffer -> memory download</dt><dd>No longer need to be flipped</dd>
   <dt>Overlarge framebuffers</dt><dd>Accounted for by flipping rendering system accounting for bufferHeight</dd>
   <dt>Texturing from framebuffer offset</dt><dd>Accounted for when blitting</dd>
   <dt>Memory -> framebuffer offset</dt><dd>Accounted for in flipped coord system on upload</dd>
   <dt>Framebuffer -> memory offset</dt><dd>Accounted for by flipping y1/y2 in y/h calculation</dd>
</dl>

I'm most concerned about the last ones.  If #4665 was fixed by this, it probably means there was a bug in that code already.  Just to make sure I'm clear, sometimes OpenGL framebuffers are reused and larger than the height we "know" they have.

Before this ended up as the ASCII art above, where X's represent non-valid pixel data.

There's two ways the coordinate system could be flipped:

```
  Original           Type 1            Type 2
------------      ------------      ------------
X XXXXXXXXXX      AAAAAAAAAA 0      XXXXXXXXXX X
X XXXXXXXXXX      BBBBBBBBBB 1      XXXXXXXXXX X
2 CCCCCCCCCC  ->  CCCCCCCCCC 2  vs  AAAAAAAAAA 0
1 BBBBBBBBBB      XXXXXXXXXX X      BBBBBBBBBB 1
0 AAAAAAAAAA      XXXXXXXXXX X      CCCCCCCCCC 2
```

I think we are using the more logical type 1 (which will avoid problems with resizing.)  As such, when flipping offsets, we need to account for this.

An example of download (glReadPixels):

```
  01234
0 ABCDE 6
1 FGHJI 5
2 JKLMN 4
3 OPQRS 3
4 TUVWX 2
- ----- 1
- ----- 0
```

OpenGL's coordinates on on the right side, sane coordinates are on the left.  When downloading 0,0 with width and height of 4, we want OpenGL row 6 first, then 5, then 4, etc.  More than that, we also have to ensure we account for the offset - OpenGL row 0 is garbage data for us in this example, since it is outside the valid pixels.

#4665 looks very much like it was previously texturing or something from a 512 tall framebuffer, and it read the wrong half, and did so even upside down.  If it was fixed now, that's great, but it seems like now other games have the reverse problem: downloads are upside down.

-[Unknown]

unknownbrackets commented on GPU/GLES/Framebuffer.cpp:
```diff
@@ -1348,7 +1359,7 @@ void FramebufferManager::BlitFramebuffer(VirtualFramebuffer *dst, int dstX, int
 		float srcW = src->bufferWidth;
 		float srcH = src->bufferHeight;
-		DrawActiveTexture(0, dstX1, dstY, w * dstXFactor, h, dst->bufferWidth, dst->bufferHeight, !flip, srcX1 / srcW, srcY / srcH, srcX2 / srcW, (srcY + h) / srcH, draw2dprogram_);
+		DrawActiveTexture(0, dstX1, dstY, w * dstXFactor, h, dst->bufferWidth, dst->bufferHeight, srcX1 / srcW, srcY / srcH, srcX2 / srcW, srcY2 / srcH, draw2dprogram_);
```
Hmm, this is a bit confusing, since the factors are fixed to 1 here.  Should probably change `dstY` to `dstY1` and `srcY` to `srcY1` to match, and either add `* dstYFactor` to `h` or remove from `w`.  Before y was always "unscaled" in this statement, although that's sorta confusing anyway.

-[Unknown]

hrydgard replied:
Hm, yeah, should clean this up. Anyway, it's in the !useBlit path so not likely to cause the remaining issues on my hardware at least..

unknownbrackets commented on GPU/GLES/Framebuffer.cpp:
```diff
@@ -1305,8 +1321,8 @@ void FramebufferManager::BlitFramebuffer(VirtualFramebuffer *dst, int dstX, int
-	int srcY2 = src->renderHeight - (h + srcY) * srcYFactor;
-	int srcY1 = srcY2 + h * srcYFactor;
+	int srcY2 = srcY * srcYFactor;
+	int srcY1 = (srcY + h) * srcYFactor;
```
These are reversed - Y1/Y2.  I think this is causing a flip.

-[Unknown]

hrydgard replied:
Oh! Good spot, I was staring at the code too long to see it... I'll try it soon.

unknownbrackets commented on GPU/GLES/StateMapping.cpp:
```diff
@@ -783,15 +793,22 @@ void TransformDrawEngine::ApplyDrawState(int prim) {
-			(regionX2 - regionX1) * renderWidthFactor,
-			(regionY2 - regionY1) * renderHeightFactor);
+		if (useBufferedRendering) {
+			// No flip needed
```
Do we still need to offset for renderHeight/bufferHeight?

If the bufferHeight is 512, but the viewport is only 272 tall, we need to make sure we're targetting rows 0-271 of the framebuffer.  Are coordinates only flipped within the viewport?

-[Unknown]

unknownbrackets commented:
Viewport y offset is off in Star Ocean 1 outside towns.  This game uses a 272 Y scale (meaning, 272 * 2 total Y viewport height.)

-[Unknown]

hrydgard committed:
Fix a Y flip typo

hrydgard commented:
Yeah, I believe the viewport clipping code in StateMapping might not be working properly. It hasn't been converted to view the world the new way...

hrydgard committed:
Make some optional arguments non-optional.

hrydgard commented:
@unknownbrackets  in "Framebuffer -> memory offset"

Where's this flipping calculation you refer to? ("Accounted for by flipping y1/y2 in y/h calculation")

unknownbrackets commented:
Oh, btw, we need to set flipped = false in all the debug buffers at the end of Framebuffer.cpp.  I can send a pull.

About the offset... when downloading:

```c++
		int byteOffset = y * vfb->fb_stride * 4;
		glReadPixels(0, y, vfb->fb_stride, h, glfmt, GL_UNSIGNED_BYTE, packed + byteOffset);
```

Right now, it offsets the memory and y1 position it downloads from.  This code was designed for flipped framebuffers, though, and so I think somewhere things need to be accounted for.

-[Unknown]

hrydgard commented:
I think that should be right, actually - seems it might have been wrong before?

hrydgard committed:
Set the flipped flag to false on debug buffers

unknownbrackets commented:
Maybe I'm confused.  This is all very annoying.

```
  01234
0 ABCDE 6
1 FGHJI 5
2 JKLMN 4
3 OPQRS 3
4 TUVWX 2
- ----- 1
- ----- 0
```

Won't glReadPixels y=0, h=4 do the following:

1. Start at OpenGL Y = 0 (right side 0).
2. Download one stride and place it in memory.
3. Advance y by 1 (OpenGL y = 1 on the right side.)
4. Download one stride again, and repeat 4 times.
5. Finalize at OpenGL Y = 3, the last row downloaded.

This means that memory will now look like this, right?

```
-----
-----
TUVWX
OPQRS
```

Or am I missing something?  Before, it was flipped, so it worked out:

```
  01234
- ----- 6
- ----- 5
4 TUVWX 4
3 OPQRS 3
2 JKLMN 2
1 FGHJI 1
0 ABCDE 0
```

-[Unknown]

hrydgard commented:
Well, now that we don't flip Y/V in various ways anymore, except for the final blit to the screen, you could  now consider 0,0 to be top left in image space, it's really arbitrary if we call positive Y "up" or "down". glReadPixel will read from increasing Y line numbers into increasing memory addresses. When you consider GL coordinates the normal way, that would be upwards, but in our point of view, it's downwards and glReadPixels should thus behave the way we want...

hrydgard committed:
Fix some typos in D3D9 blits (removed parameter).

Add some missing projection matrix flips (flipped viewport emulation)

Fix issue with DrawPixels in non-buffered in D3D9

unknownbrackets commented:
Okay, sorry, I had it backwards.  Thanks.

Well, the only thing I'm seeing wrong at this point is the viewport in Star Ocean 1.  Grand Knights History and 3rd Birthday work with the other fixes so far.

-[Unknown]

unknownbrackets committed:
Switch viewport overage to the new world of sanity.

LunaMoo commented:
Should this branch help with #8016 at this point? Since there doesn't seem to be any difference for that.

 Great work on figuring out all those problems through, seems like it's not breaking anything anymore while taking care of another ancient issue:).
 Guessing will have to rebase my small display replacement again after this:3.

hrydgard commented:
@LunaMoo  No you won't, I'm going to merge this first and rebase that branch myself, don't worry.

LunaMoo commented:
Oh I'm not worried about that:3, I'm actually glad I had to use rebase a few times since I finally learned how and it's pretty easy even when there are conflicts to resolve:). Git is kind of overwhelming without some practice.

hrydgard commented:
Oh and I forgot to answer, yeah, it will help with #8016. I will commit the fixes to that after merging this.

hrydgard commented:
Last commit broke Wipeout in GL buffered, will need to figure out that before merge.

BTW does Star Ocean work in cities with all the stencil stuff?

hrydgard committed:
unknownbrackets' fixes for last commit
